pkg/concurrency/stopper.go

```diff
@@ -1,35 +1,115 @@
 package concurrency
 
-// NewStopper creates a new Stopper interface
+// NewStopper creates a new Stopper for arranging a graceful shutdown.
 func NewStopper() Stopper {
-	return Stopper{
-		stop:    NewSignal(),
-		stopped: NewSignal(),
+	return &stopperImpl{
+		stop:    NewErrorSignal(),
+		stopped: NewErrorSignal(),
 	}
 }
 
-// Stopper is an object that encapsulates both stop and stopped signals
-type Stopper struct {
-	stop    Signal
-	stopped Signal
+// Stopper encapsulates stop and stopped signals for arranging a graceful shutdown sequence for goroutines (and async
+// processing in general). These are referred as "stoppable" goroutines.
+type Stopper interface {
+	// Client returns an interface that should be used outside the goroutine to request a graceful shutdown and check
+	// the shutdown status.
+	Client() StopperClient
+
+	// Flow returns an interface that should be used in stoppable goroutine control flow to check and modify the state
+	// of Stopper. Flow should not be used outside the stoppable goroutine.
+	Flow() StopperFlow
+
+	// LowLevel allows to interact with the stopper at lower level. Callers should avoid using it unless there's no
+	// better alternative.
+	LowLevel() StopperLowLevel
+}
+
+// StopperClient represents an interface for an external client (external to a stoppable goroutine) to request the
+// stoppable goroutine to gracefully shutdown and to check the shutdown status.
+type StopperClient interface {
+	// Stop requests stoppable goroutine to shut down (or stop).
+	Stop()
+
+	// Stopped returns ReadOnlyErrorSignal that can be used for waiting until the stoppable goroutine reports that it
+	// finally stopped.
+	Stopped() ReadOnlyErrorSignal
 }
 
-// Stop signals the internal stop signal and effectively causes a push to StopDone
-func (s *Stopper) Stop() {
+// StopperFlow represents an interface that should be used by a stoppable goroutine to check and report its status of a
+// graceful shutdown.
```

Comment:
If I understood your idea correctly, the `Flow` construct should be used by the sensor components if a parent (or any other component) requests a shutdown?

Reply:
Flow is to be used by goroutine. Flow allows to check if shutdown was requested and allows to report that the shutdown was completed.
Client is to be used outside of goroutine. Client allows to request shutdown and allows to way for the shutdown to complete.
Makes sense?

Reply:
Ok, after reading the code again a few times I think it's making more sense now. Thanks.